src/test/shared/ui/pickerPrompter.test.ts

```diff
@@ -277,124 +315,104 @@ describe('QuickPickPrompter', function () {
 
             await testPrompter.loadItems(newItems)
             await testPrompter.loadItems(newItemsPromise)
         })
+        tester.acceptItem(filterBoxInput.label)
 
-        assert.strictEqual(await loadAndPrompt(), Number(input))
+        await tester.result(Number(input))
     })
 
     it('can accept custom input as a last response', async function () {
         const input = '123'
 
-        testPrompter.onDidShow(() => {
-            picker.onDidChangeActive(active => {
-                if (active[0]?.description !== undefined) {
-                    picker.selectedItems = [active[0]]
-                    picker.fireOnDidAccept()
-                }
-            })
-
-            testPrompter.recentItem = { data: CUSTOM_USER_INPUT, description: input } as any
-            picker.fireOnDidChangeValue(input)
+        tester.addCallback(prompter => {
+            prompter.recentItem = { data: CUSTOM_USER_INPUT, description: input } as any
+            picker.fireOnDidChangeValue(picker.value)
         })
+        tester.acceptItem(filterBoxInput.label)
 
-        assert.strictEqual(await loadAndPrompt(), Number(input))
+        await tester.result(Number(input))
     })
 
     it('validates the custom input', async function () {
         const input = 'not a number'
 
-        testPrompter.onDidShow(() => {
-            const disposable = picker.onDidChangeActive(items => {
-                if (
-                    items[0]?.description === input &&
-                    items[0]?.detail?.includes('NaN') &&
-                    items[0]?.invalidSelection
-                ) {
-                    picker.onDidChangeActive(items => {
-                        if (items.length > 0) {
-                            picker.selectedItems = [picker.items[0]]
-                        }
-                    })
-                    picker.selectedItems = [picker.items[0]]
-                    disposable.dispose()
-                    picker.value = ''
-                    picker.fireOnDidChangeValue('')
-                }
-            })
+        tester = createQuickPickTester(testPrompter, { forceEmits: true })
+        tester.setFilter(input)
+        tester.acceptItem(filterBoxInput.label)
+        tester.setFilter(undefined)
+        tester.acceptItem(testItems[0])
+
+        await tester.result(testItems[0].data)
+    })
+
+    it('can handle async validation', async function () {
+        const input = 'not a number'
+        const validator = async (val: string) => {
+            await new Promise(r => setImmediate(r))
+            return filterBoxInput.validator(val)
+        }
+        const filterBox = { ...filterBoxInput, validator }
+
+        const newPicker = vscode.window.createQuickPick() as DataQuickPick<number>
+        testPrompter = new FilterBoxQuickPickPrompter(newPicker, { filterBoxInput: filterBox })
+        tester = createQuickPickTester(testPrompter, { forceEmits: true })
```

Comment:
Setting the `forceEmits` option will trigger the event emitter for certain actions rather than letting VS Code fire the event. The reason for this is that VS Code will trigger the event 'at some point', though this leads to flaky tests. This flag could potentially lead to a deviation in test behavior vs. actual which is why it is false by default.